stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+
+![Linked List Overview](img/linked_list_overview.png)
+
+### Need of Linked Lists
+With linked lists, you never have to move your items: linked lists have the ability to break the chain and rejoin it. In other words, if you wanted to put a new element between Node №1 and Node №2, the steps would be:
+- Create a new Node and allocate memory to it;
+
+![Linked List Need №1](img/linked_list_need_1.png)
+
+- Add a new "data" value;
+
+![Linked List Need №2](img/linked_list_need_2.png)
+
+- Point its "next" pointer to Node №2;
+
+![Linked List Need №3](img/linked_list_need_3.png)
+
+- Change the "next" pointer of Node №1 to a newly created Node.
+
+![Linked List Need №4](img/linked_list_need_4.png)
+
+Doing something similar in an array would have required shifting the positions of all the subsequent elements.
+
+### Pros and Cons of Arrays and Linked Lists
+Which data structure should we use: arrays or linked lists? It depends only on a particular use case and goals, which we intend to resolve. In order to make a balanced decision, we should know pros and cons of these data structures.
+
+#### Situation №1: Read of the last element
+A read operation is heavily tied to type of access. There are two types of access to elements of a data structure: random access and sequential access.
+Sequential access means reading all elements one by one starting from the 1st element. Random access means that we can reach a needed element directly. Linked lists can only perform sequential access, when arrays can do both sequential and random types of access.
+
+Reading of the last element may be a quite heavy operation for a linked list: we can't just read it, because you don't know what address it has. We have to go to Node №1 to get the address for Node №2, then until we get to the last Node. Thus, linked lists are great if we are going to read all the Nodes one at a time, and they are terrible, if we are planning to jump around its Nodes: first, access the last element, then the 3rd element, then the last once again.
+
+Regarding arrays, we know the address for each element, therefore arrays are great if we want to read random elements, because you can look up any element in your array instantly.
+
+#### Situation №2: Insert a new element
+Let's suppose a situation when we should insert a new element in a collection. As concerns linked lists, this operation will be very cheap, because we should just make a previous node point at a newly created node. An example of insertion can be found [here](#need-of-linked-lists).
```

Comment:
this is also a havy operation for a LinkedList if we insert to the middle(because we need to get through this sequential search to find the required element)

Reply:
the same with deletions